### ISSUE TITLE
In-built request timeout functionality

### ISSUE THREAD

**Issue by Renegade334** [reactphp/http-client#28]
Currently, request timeouts can be implemented somewhat messily by specifying loop timers and attempting to manually close the request. They should ideally be in-built.

**Comment by cboden** [comment]
This feature has been started in the underlying library reactphp/socket-client#10. One it's implemented there it should be simple enough to add here.

**Comment by clue** [comment]
:+1: on supporting a timeout setting.

Though I'm not sure what the API could look like…
Also, *when* should we actually time out (connection timeout, whole request timeout, idle timeout etc…)?

I think it makes sense to add an API similar to https://github.com/reactphp/socket-client/pull/10. Any thoughts on this @Renegade334?

**Comment by Renegade334** [comment]
Clients like cURL support both a connection timeout and a total timeout. Both would be useful, IMO. The latter would be a client-level feature, I guess, whereas the former would be socket-level.

```
renegade@mercury:~$ curl --connect-timeout 1 http://10.255.255.1/
curl: (28) Connection timed out after 1001 milliseconds
renegade@mercury:~$ netcat -l 3456 &> /dev/null & curl --max-time 1 http://localhost:3456/
curl: (28) Operation timed out after 1000 milliseconds with 0 bytes received
```

**Comment by WyriHaximus** [comment]
@Renegade334 yes that makes sense. We could even go for 3 time outs:
* Connect timeout, how long we wait for the connection to be created, done in `socket-client`
* Response timeout, how long a server can take before responding
* Response body timeout, how long it can take before the entire response body is downloaded

**Comment by clue** [comment]
FWIW: This is likely related to #36. Possible implementation approach could be to just `close()` the request once the timeout timer fires.

**Comment by clue** [comment]
I'm closing this as it is already supported by the new HTTP client that has been merged into `react/http` via reactphp/http#368 :shipit: You should upgrade as per #152 :+1:
